pyi.py

```diff
@@ -425,6 +427,77 @@ def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
     def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef) -> None:
         self._visit_function(node)
 
+    def _Y019_error(
+        self, node: ast.FunctionDef | ast.AsyncFunctionDef, typevar_name: str
+    ) -> None:
+        error_message = Y019.format(typevar_name=typevar_name)
+
+        if sys.version_info >= (3, 9):
+            cleaned_method = deepcopy(node)
+            cleaned_method.decorator_list.clear()
+            new_syntax = re.sub(
+                fr"(\W){typevar_name}(\W)", r"\1Self\2", ast.unparse(cleaned_method)
+            )
+            new_syntax = re.sub(r"\s+", " ", new_syntax)
+            error_message += f', e.g. "{new_syntax}"'
+
+        # pass the node for the first argument to `self.error`,
+        # rather than the function node,
+        # as linenos differ in Python <3.9 for decorated functions
+        self.error(node.args.args[0], error_message)
+
+    def _check_instance_method_for_bad_typevars(
+        self,
+        *,
+        method: ast.FunctionDef | ast.AsyncFunctionDef,
+        first_arg_annotation: ast.Name | ast.Subscript,
+        return_annotation: ast.Name,
+    ) -> None:
+        if not isinstance(first_arg_annotation, ast.Name):
+            return
+
+        if first_arg_annotation.id != return_annotation.id:
+            return
+
+        arg1_annotation_name = first_arg_annotation.id
+
+        if arg1_annotation_name.startswith("_"):
+            self._Y019_error(method, arg1_annotation_name)
+
+    def _check_class_method_for_bad_typevars(
+        self,
+        *,
+        method: ast.FunctionDef | ast.AsyncFunctionDef,
+        first_arg_annotation: ast.Name | ast.Subscript,
+        return_annotation: ast.Name,
+    ) -> None:
+        if not isinstance(first_arg_annotation, ast.Subscript):
+            return
+
+        cls_typevar: str
+
+        # see comment in visit_Subscript
```

Comment:
I think this comment is going to get outdated as code is being moved around. I think we should just add a new utility method that returns `node.slice.value` or `node.slice` depending on Python version. After this PR, there will be at least 3 different places where it would be useful.

Reply:
Yes, good shout.

Reply:
#92 does already do something similar.

Reply:
Let's merge #92 first, then